libtsuba/include/tsuba/Cache.h

```diff
@@ -0,0 +1,182 @@
+#ifndef KATANA_LIBTSUBA_TSUBA_CACHE_H_
+#define KATANA_LIBTSUBA_TSUBA_CACHE_H_
+
+// This is single threaded only, it is not intended to store large objects,
+// but rather metadata (e.g., a shared_ptr to a property column).
+
+// The problem witchel had implementing a multi-threaded version using
+// parallel-hashmap is a lock ordering problem.  parallel-hashmap 1.33 allows
+// execution of code with the write lock held, only when modifying, not when adding an
+// element.  We have to modify the LRU list when eviciting an element, so the natural
+// lock ordering is parallel-hashmap write lock, then list lock.  But without a way to
+// execute insert code with the parallel-hashmap write lock held, it seemed like there
+// would be some form of race condition.
+
+#include <list>
+#include <optional>
+#include <string>
+
+#include <boost/container_hash/hash.hpp>
+
+#include "katana/ArrowInterchange.h"
+#include "katana/ConcurrentMap.h"
+#include "katana/Logging.h"
+
+namespace tsuba {
+
+struct KATANA_EXPORT KeyBase {
+  enum class NodeEdge { kNode, kEdge };
+  bool operator==(const KeyBase& o) const {
+    return node_edge_ == o.node_edge_ && name_ == o.name_;
+  }
+
+  NodeEdge node_edge_;
+  std::string name_;
+
+  struct Hash {
+    std::size_t operator()(const KeyBase& key) const {
+      using boost::hash_combine;
+      using boost::hash_value;
+
+      std::size_t seed = 0;
+      hash_combine(seed, hash_value(key.node_edge_));
+      hash_combine(seed, hash_value(key.name_));
+
+      // Return the result.
+      return seed;
+    }
+  };
+};
+
+template <typename Value>
+class KATANA_EXPORT Cache {
+public:
+  using ListType = std::list<KeyBase>;
+  struct MapValue {
```

Comment:
It doesn't look like `MapValue` or `MapType` need to be public. Can we make them private instead?